amp-cli: rewrite using writeShellApplication

ghuntley opened the pull request:
👋🏻 On the core team that produces [Amp](https://ampcode.com). Our update cadence is at a minimum of four times a day, so pinning via `buildNpmPackage` is unsuitable. Rewrote as `writeShellApplication` with `nodejs_22` runtime dependency which executes `npx --yes @sourcegraph/amp` instead.


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ghuntley committed:
amp-cli: rewrite using writeShellApplication

- Replace buildNpmPackage with writeShellApplication
- Use nodejs_22 instead of nodejs_20 (required by amp CLI)
- Execute "npx --yes @sourcegraph/amp" to pass through all arguments
- Remove unnecessary package-lock.json and update.sh files
- Add ghuntley as maintainer

Amp-Thread: https://localhost:7002/threads/T-2017874a-39cf-4ae6-9f4a-2a4d43648805
Co-authored-by: Amp <amp@ampcode.com>

ghuntley committed:
Update pkgs/by-name/am/amp-cli/package.nix

ghuntley commented on pkgs/by-name/am/amp-cli/package.nix:
```diff
@@ -1,80 +1,31 @@
+      runtimeInputs = [ nodejs_22 ];
+      text = ''exec npx --yes @sourcegraph/amp "$@"'';
+    });
+    command = "HOME=$(mktemp -d) amp-cli --version";
```
```suggestion
    command = "HOME=$(mktemp -d) amp --version";
```

ghuntley committed:
Update pkgs/by-name/am/amp-cli/package.nix

ghuntley committed:
Update pkgs/by-name/am/amp-cli/package.nix

ghuntley commented on pkgs/by-name/am/amp-cli/package.nix:
```diff
@@ -1,80 +1,31 @@
-    "--no-fund"
-    "--ignore-scripts"
+writeShellApplication {
+  name = "amp-cli";
```
```suggestion
  name = "amp";
```

ghuntley committed:
Update pkgs/by-name/am/amp-cli/package.nix

theCapypara commented:
Is there any example of a package like this on nixpkgs? I personally don't think it's a good idea to have a wrapper program that just executes arbitrary code from the internet in nixpkgs.

jvmncs commented:
Maybe this can be split into two packages? A slower cadence, buildNpmPackage-based package, along with an amp-bin that moves faster? There's some precedent for this elsewhere in nixpkgs (although those are usually versioned explicitly, so not sure it's exactly the same).

ghuntley commented:
> A slower cadence, buildNpmPackage-based package
Unfortunately not, our internal APIs are constantly evolving and having Nix users use an older one will result in pain for the Nixpkgs project and also for us on support.

ghuntley committed:
amp-cli: rename to ampcode with backward compatibility alias

- Renamed package from amp-cli to ampcode
- Added backward compatibility alias: amp-cli = ampcode
- Shell script remains named "amp" as requested
- Maintains all existing functionality and metadata

Amp-Thread: https://ampcode.com/threads/T-72f14e4d-dc4e-4fc1-8cef-aec1772602a1
Co-authored-by: Amp <amp@ampcode.com>

ghuntley reviewed (commented):
Changes:
- renamed from `amp-cli` to `ampcode` (and implemented alias)
- ran nixpkgs formatter

Good to go if CI passes.

jvmncs commented:
+1 on a merge

adamcstephens commented:
I also just discovered https://github.com/numtide/llm-agents.nix which includes auto updated amp cli.
